File: src/main/java/javaclasses/LinkedList/RotateList.java
/*
 * * Rotate List.java
 *  * Created by Rafsan Ahmad on 10/25/21, 9:49 PM
 *  * Copyright (c) 2021 . All rights reserved.
 *
 */

package javaclasses.LinkedList;

public class RotateList {
    //Leetcode 61
    /*Given the head of a linked list, rotate the list to the right by k places.
    * Input: head = [1,2,3,4,5], k = 2
    Output: [4,5,1,2,3]*/

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return null;
        ListNode copyHead = head;
        int len = 1;
        while (copyHead.next != null) {
            copyHead = copyHead.next;
            len++;
        }
        copyHead.next = head;
        for (int i = len - k % len; i > 1; i--) {
            head = head.next;
        }
        copyHead = head.next;
        head.next = null;
        return copyHead;
    }

    public void printList(ListNode node) {
        while (node != null) {
            System.out.print(node.val + " ");
            node = node.next;
        }
    }

    public static void main(String[] args) {
        ListNode node = new ListNode(1);
        node.next = new ListNode(2);
        node.next.next = new ListNode(3);
        node.next.next.next = new ListNode(4);
        node.next.next.next.next = new ListNode(5);
        RotateList rt = new RotateList();
        ListNode result = rt.rotateRight(node, 2);
        rt.printList(result);
    }
}

File: src/main/java/kotlinclasses/LinkedList/RotateList.kt
/*
 * *
 *  * Rotate List.kt
 *  * Created by Rafsan Ahmad on 7/8/25, 8:18PM
 *  * Copyright (c) 2025 . All rights reserved.
 *  *
 *
 */

package kotlinclasses.LinkedList

class RotateList {
    //https://leetcode.com/problems/rotate-list/description/
    /*Given the head of a linked list, rotate the list to the right by k places.

Example 1:
Input: head = [1,2,3,4,5], k = 2
Output: [4,5,1,2,3]

Example 2:
Input: head = [0,1,2], k = 4
Output: [2,0,1]


Constraints:
The number of nodes in the list is in the range [0, 500].
-100 <= Node.val <= 100
0 <= k <= 2 * 10^9*/

    fun rotateRight(head: ListNode?, k: Int): ListNode? {
        if (head == null || k == 0) return head
        var len = 0
        val dummy = ListNode(0)
        dummy.next = head
        var temp = head

        //Step 1: Find Length
        while (temp != null) {
            len++
            temp = temp.next
        }

        temp = head
        var prev: ListNode? = null
        if (k % len == 0) return head
        var rotatePoint = len - k % len

        //Step:2 - Find rotate point
        //1, 2, 3, 4, 5, k=2, rotatePoint = 4
        while (rotatePoint > 0) {
            prev = temp
            temp = temp?.next
            rotatePoint--
        }

        //Step:3 - Connect head and set null for last element
        //1, 2, 3, 4, 5, k=2
        //1-2-3-Null
        prev?.next = null
        dummy.next = temp

        //Step:4 - Go until last elemnt in list and connect head
        //1, 2, 3, 4, 5 -> 4,5-1,2,3
        while (temp?.next != null) {
            temp = temp.next
        }
        temp?.next = head
        return dummy.next
    }
}

fun main() {
    val obj = RotateList()
    val input = listOf(1, 2, 3, 4, 5)
    val head = buildList(input)
    val result = obj.rotateRight(head, 2)
    println(toList(result))

    val input2 = listOf(0, 1, 2)
    val head2 = buildList(input2)
    val result2 = obj.rotateRight(head2, 4)
    println(toList(result2))
}